### PR TITLE
Refactor: FSJobCatalog V1 & V2

### PR DESCRIPTION
**Pull request by autumnust**
Verison 2 of FSjobcatalog based on refactored SchedulerUtils.(Moving some of the functions into FSjobCatalogLoadingHelper )

Close the original PR. 

This one is cleaner to review. Please take a look at @ibuenros , Thanks.

### CODE REVIEW

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+  /**
+   * The expose of observer is used for testing purpose, so that
+   * the checkAndNotify method can be revoked manually, instead of waiting for
+   * the scheduling timing.
+   * @param jobConfig The same as general constructor.
+   * @param observer The user-initialized observer.
+   * @throws Exception
+   */
+  public FSJobCatalog(Properties jobConfig, PathAlterationObserver observer)
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java]: Can you make this `protected` and add the annotation `@VisibleForTesting`?

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+  }
+
+  /**
+   * Initialize the JobCatalog, fetch all jobs in jobConfDirPath.
+   * @param jobConfig
+   * @throws Exception
+   */
+  public FSJobCatalog(Properties jobConfig)
+      throws Exception {
+    this(jobConfig, null);
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java]: shouldn't the observer be non-null?

File: gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java [hunk 3]
```diff
@@ -113,7 +113,7 @@
   public final Set<String> jobConfigFileExtensions;
 
   // A monitor for changes to job conf files for general FS
-  public final PathAlterationMonitor pathAlterationMonitor;
+  public final PathAlterationDetector _pathAlterationDetector;
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java]: please undo this change

File: gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogJobSpecHelper.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  /**
+   * On receiving a JobSpec object, materialized it into a real file into non-volatile storage layer.
+   * @param jobConfDirPath The configuration root directory path.
+   * @param jobSpec a {@link JobSpec}
+   */
+  public static void materializeJobSpec(Path jobConfDirPath, JobSpec jobSpec)
+      throws IOException {
+    Path path = new Path(jobConfDirPath, jobSpec.getUri().toString());
+    try (FileSystem fs = path.getFileSystem(new Configuration())) {
+      Properties props = ConfigUtils.configToProperties(jobSpec.getConfig());
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogJobSpecHelper.java]: No need, you can write `Config` directly.

File: gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogJobSpecHelper.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  private static final String VERSION_KEY_IN_JOBSPEC = "version";
+
+  /**
+   * On receiving a JobSpec object, materialized it into a real file into non-volatile storage layer.
+   * @param jobConfDirPath The configuration root directory path.
+   * @param jobSpec a {@link JobSpec}
+   */
+  public static void materializeJobSpec(Path jobConfDirPath, JobSpec jobSpec)
+      throws IOException {
+    Path path = new Path(jobConfDirPath, jobSpec.getUri().toString());
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogJobSpecHelper.java]: Please use `PathUtils.mergePaths` instead, as this will not work when the `JobSpec` uri is absolute

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java (hunk 1 above)
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java]: Also, can you change all instances of `Properties` to `Config`?

File: gobblin-utility/src/main/java/gobblin/util/PathUtils.java [hunk 6]
```diff
@@ -31,7 +31,7 @@
   public static final Pattern GLOB_TOKENS = Pattern.compile("[,\\?\\*\\[\\{]");
 
   public static Path mergePaths(Path path1, Path path2) {
-    String path2Str = path2.toUri().getPath();
+    String path2Str = "/"+path2.toUri().getPath();
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/PathUtils.java]: Do not change this method.

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+  /**
+   * Initialize the JobCatalog, fetch all jobs in jobConfDirPath.
+   * @param sysConfig
+   * @throws Exception
+   */
+  public FSJobCatalog(Config sysConfig)
+      throws Exception {
+    this(sysConfig, null);
+    /* Automatically shceudle the checkAndNotify method using detector's threadpool, in general usage. */
+    this.pathAlterationDetector.start();
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java]: Doesn't this happen in the superclass?

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+  /**
+   * Listeners of JobCatalog, Like GobblinInstaceDriver.
+   * Distinguished it from JobSpec Listeners.
+   */
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(FSJobCatalog.class);
+
+  // A monitor for changes to job conf files for general FS
+  // This embedded monitor is monitoring job configuration files instead of JobSpec Object.
+  private final PathAlterationDetector pathAlterationDetector;
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java]: Why do you need this here instead of superclass?

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+  private final PathAlterationObserver observer;
+
+  /**
+   * Initialize the JobCatalog, fetch all jobs in jobConfDirPath.
+   * @param sysConfig
+   * @throws Exception
+   */
+  public FSJobCatalog(Config sysConfig)
+      throws Exception {
+    this(sysConfig, null);
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java]: Shouldn't the observer be not null?

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+  @Override
+  public synchronized void put(JobSpec jobSpec) {
+    Preconditions.checkNotNull(jobSpec);
+    try {
+      String srcFileName = jobSpec.getUri().toString();
+      Path tmpPath = PathUtils.mergePaths(this.jobConfDirPath, new Path(jobSpec.getUri().toString()));
+      if (fs.exists(tmpPath)) {
+        LOGGER.info("The job with URI[" + tmpPath + "] has been added before, will cover the original one.");
+
+        FSJobCatalogHelper.materializedJobSpec(new Path(srcFileName), this.jobConfDirPath, jobSpec, true);
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java]: The last argument (`true`) should be a call to a method `e.g. shouldLoadGlobalConf()`, so that the mutable subclass can override this to `false`.

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/ImmutableFSJobCatalog.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  /**
+   * Fetch all the job files under the jobConfDirPath
+   * @return A collection of JobSpec
+   */
+  @Override
+  public synchronized List<JobSpec> getJobs() {
+    return Lists.transform(
+        Lists.newArrayList(loader.loadPullFilesRecursively(loader.getRootDirectory(), this.sysConfig, true)),
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/ImmutableFSJobCatalog.java]: The last argument (true) should be a call to a method e.g. shouldLoadGlobalConf(), so that the mutable subclass can override this to false.

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/ImmutableFSJobCatalog.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+   * Fetch single job file based on its URI,
+   * return null requested URI not existed
+   * @param uri The relative Path to the target job configuration.
+   * @return
+   */
+  @Override
+  public synchronized JobSpec getJobSpec(URI uri)
+      throws JobSpecNotFoundException {
+    try {
+      Path targetJobSpecFullPath = PathUtils.mergePaths(this.jobConfDirPath, new Path(uri));
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/ImmutableFSJobCatalog.java]: Please refactor this into a separate method, e.g. `getPathForUri`

File: gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogHelper.java [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+    private Path jobConfigDirPath;
+
+    public JobSpecConverter(Path jobConfigDirPath) {
+      this.jobConfigDirPath = jobConfigDirPath;
+    }
+
+    @Nullable
+    @Override
+    public JobSpec apply(@Nullable Config inputConfig) {
+      return jobSpecConverter(inputConfig, jobConfigDirPath);
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogHelper.java]: There's no point having this logic in a different method, just move it here.

File: gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogHelper.java [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+    observer.addListener(listener);
+    detector.addObserver(observer);
+  }
+
+  /**
+   * @param jobConfigDirPath The root directory of job configuration files.
+   * @param path Unknown for whether it is full path or not. Will need to check this first.
+   * @return Resovled full path.
+   */
+  public static Path resovledRelativePath(Path jobConfigDirPath, Path path) {
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogHelper.java]: why do you need this? There should be a clear division: user generated paths are always "relative", while internal paths always start with `jobConfigDirPath`.

File: gobblin-utility/src/test/java/gobblin/util/SchedulerUtilsTest.java [hunk 15]
```diff
@@ -306,5 +307,16 @@ private Properties getJobConfigForFile(List<Properties> jobConfigs, String fileN
     }
     return null;
   }
+
+  /**
+   * Only for print the attributes name for loaded job configuration files.
+   * For inspection purpose.
+   */
+  private void printAttributeName(Properties jobProps, int position ) {
```
Comment (ibuenros) [inline gobblin-utility/src/test/java/gobblin/util/SchedulerUtilsTest.java]: Is this needed? If so, you should use logger not stderr

File: gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogHelper.java (hunk 13 above)
Comment (autumnust) [inline gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogHelper.java]: The function interface is only used for transforming each elements in list into the type that we expected. But this method has other usage, like in `FSJobCatalog:170` and `ImmutableFSJobCatalog:82`

File: gobblin-utility/src/test/java/gobblin/util/SchedulerUtilsTest.java (hunk 15 above)
Comment (autumnust) [inline gobblin-utility/src/test/java/gobblin/util/SchedulerUtilsTest.java]: No needed. Deleted.

File: gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogHelper.java (hunk 13 above)
Comment (autumnust) [inline gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogHelper.java]: Plus this converter method is believed to be more flexible to use than function interface.

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java (hunk 9 above)
Comment (autumnust) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java]: This null issue is addressed by `addPathAlterationObserver` method, which contains `Optional<PathAlterationObserver>` as a parameter. 

The initialization of this parameter is: `Optional<PathAlterationObserver> observerOptional = Optional.fromNullable(this.observer);` 

It could be null here and leave it to the lower layer methods to handle.

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/ImmutableFSJobCatalog.java [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+  private final FSJobCatalogHelper.JobSpecConverter converter;
+
+  // A monitor for changes to job conf files for general FS
+  // This embedded monitor is monitoring job configuration files instead of JobSpec Object.
+  protected final PathAlterationDetector pathAlterationDetector;
+  protected final PathAlterationObserver observer;
+
+  public ImmutableFSJobCatalog(Config sysConfig)
+      throws Exception {
+    this(sysConfig, null);
```
Comment (autumnust) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/ImmutableFSJobCatalog.java]: This null issue is addressed by `addPathAlterationObserver` method, which contains `Optional<PathAlterationObserver>` as a parameter. 

The initialization of this parameter, inside the `JobCatalog` constructor is: `Optional<PathAlterationObserver> observerOptional = Optional.fromNullable(this.observer);` 

It could be null here and leave it to the lower layer methods to handle.

**Comment by ibuenros** [comment]
LGTM
